<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>src.homology</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="src-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://web.cs.dal.ca/~safatli/">LabBlouinTools</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="src-module.html">Package&nbsp;src</a> ::
        Module&nbsp;homology
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="src.homology-module.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== MODULE DESCRIPTION ==================== -->
<h1 class="epydoc">Module homology</h1><p class="nomargin-top"><span class="codelink"><a href="src.homology-pysrc.html">source&nbsp;code</a></span></p>
<!-- ==================== CLASSES ==================== -->
<a name="section-Classes"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Classes</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Classes"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="src.homology.manualModeller-class.html" class="summary-name">manualModeller</a><br />
      Set up and run Modeller using a given folder of PDB files and a 
        target FASTA sequence.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="src.homology.fasta-class.html" class="summary-name">fasta</a><br />
      Open, read, and organize a FASTA file as an object.
    </td>
  </tr>
</table>
<!-- ==================== FUNCTIONS ==================== -->
<a name="section-Functions"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Functions</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Functions"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#cleanFastaFile" class="summary-sig-name">cleanFastaFile</a>(<span class="summary-sig-arg">fname</span>,
        <span class="summary-sig-arg">targetfolder</span>=<span class="summary-sig-default">''</span>)</span><br />
      Removes gaps from target FASTA file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#cleanFastaFile">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="writeModellerPIR"></a><span class="summary-sig-name">writeModellerPIR</span>(<span class="summary-sig-arg">pdb_files</span>,
        <span class="summary-sig-arg">seq</span>,
        <span class="summary-sig-arg">dest_file</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#writeModellerPIR">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="writeAlignmentFromFasta"></a><span class="summary-sig-name">writeAlignmentFromFasta</span>(<span class="summary-sig-arg">fasta_file</span>,
        <span class="summary-sig-arg">dest_file</span>)</span><br />
      Reads a FASTA file and outputs it to a Clustal-like alignment.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#writeAlignmentFromFasta">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="writeSOAPSSAlignment"></a><span class="summary-sig-name">writeSOAPSSAlignment</span>(<span class="summary-sig-arg">fasta_file</span>,
        <span class="summary-sig-arg">pdb1</span>,
        <span class="summary-sig-arg">pdb2</span>,
        <span class="summary-sig-arg">dest_file</span>)</span><br />
      Reads a FASTA file and outputs to a SOAPSS alignment file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#writeSOAPSSAlignment">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="writeFirstFromFasta"></a><span class="summary-sig-name">writeFirstFromFasta</span>(<span class="summary-sig-arg">fasta_file</span>,
        <span class="summary-sig-arg">dest_file</span>)</span><br />
      Extracts first sequence from a FASTA and writes to a new one.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#writeFirstFromFasta">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#extractRandomFasta" class="summary-sig-name">extractRandomFasta</a>(<span class="summary-sig-arg">fasta_file</span>,
        <span class="summary-sig-arg">numRandom</span>,
        <span class="summary-sig-arg">dest_file</span>)</span><br />
      Extracts a random number of sequences from a FASTA file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#extractRandomFasta">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="completePDB"></a><span class="summary-sig-name">completePDB</span>(<span class="summary-sig-arg">pdbin</span>,
        <span class="summary-sig-arg">pdbout</span>)</span><br />
      Given a PDB, clean/complete the PDB using Modeller's complete_pdb 
      function.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#completePDB">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="getRandomPDBFragment"></a><span class="summary-sig-name">getRandomPDBFragment</span>(<span class="summary-sig-arg">pdb</span>,
        <span class="summary-sig-arg">k</span>=<span class="summary-sig-default">100</span>)</span><br />
      Given a PDB, get a random fragment of length k.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#getRandomPDBFragment">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="getFirstResidueNumber"></a><span class="summary-sig-name">getFirstResidueNumber</span>(<span class="summary-sig-arg">pdb</span>)</span><br />
      Gets the FASTA residue number of the first ATOM in a PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#getFirstResidueNumber">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#checkForFileCollision" class="summary-sig-name">checkForFileCollision</a>(<span class="summary-sig-arg">path</span>,
        <span class="summary-sig-arg">ext</span>)</span><br />
      Ensures a file is not being overwritten.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#checkForFileCollision">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#system" class="summary-sig-name">system</a>(<span class="summary-sig-arg">instruction</span>)</span><br />
      Executes a system command line instruction.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#system">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="cleanModellerFolder"></a><span class="summary-sig-name">cleanModellerFolder</span>(<span class="summary-sig-arg">path</span>)</span><br />
      Moves intermediate PDBs to another subfolder in the Modeller folder 
      output from Biskit.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#cleanModellerFolder">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="getModellerPDB"></a><span class="summary-sig-name">getModellerPDB</span>(<span class="summary-sig-arg">workflowfolder</span>,
        <span class="summary-sig-arg">targetfile</span>)</span><br />
      Copies the model_00.pdb file from Modeller if present in given 
      workflow directory from homologyWorkflow.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#getModellerPDB">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="compareStructures"></a><span class="summary-sig-name">compareStructures</span>(<span class="summary-sig-arg">foldername1</span>,
        <span class="summary-sig-arg">foldername2</span>,
        <span class="summary-sig-arg">outpath</span>,
        <span class="summary-sig-arg">verbose</span>=<span class="summary-sig-default">False</span>)</span><br />
      Given 2 folders with PDB files, calculates the RMSDs with all 
      possible pairwise combinations of PDBs; writes to file located at 
      path determined by outpath.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#compareStructures">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="startPymol"></a><span class="summary-sig-name">startPymol</span>()</span><br />
      Ensures Pymol has been launched.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#startPymol">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#getAverageCorrelationCoefficient" class="summary-sig-name">getAverageCorrelationCoefficient</a>(<span class="summary-sig-arg">len_struct</span>)</span><br />
      Gets the average correlation coefficient for randomly related 
      structure pairs given the length of a structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#getAverageCorrelationCoefficient">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#radiusOfGyration" class="summary-sig-name">radiusOfGyration</a>(<span class="summary-sig-arg">pdbFile</span>,
        <span class="summary-sig-arg">chain</span>,
        <span class="summary-sig-arg">pdb</span>=<span class="summary-sig-default">None</span>)</span><br />
      Returns the approximate radius of gyration of a given PDB molecule.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#radiusOfGyration">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#tmscore" class="summary-sig-name">tmscore</a>(<span class="summary-sig-arg">alignPDB</span>,
        <span class="summary-sig-arg">alignFASTA</span>)</span><br />
      Returns the TMscore and associated P-value.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#tmscore">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#rrmsd" class="summary-sig-name">rrmsd</a>(<span class="summary-sig-arg">alignPDB</span>,
        <span class="summary-sig-arg">alignFASTA</span>,
        <span class="summary-sig-arg">rmsd</span>=<span class="summary-sig-default">False</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>)</span><br />
      Returns the rRMSD of an aligned pairwise PDB and FASTA file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#rrmsd">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.homology-module.html#rmsd" class="summary-sig-name">rmsd</a>(<span class="summary-sig-arg">pdbFile1</span>,
        <span class="summary-sig-arg">pdbFile2</span>)</span><br />
      Returns the RMSD of two PDB files.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.homology-pysrc.html#rmsd">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
</table>
<!-- ==================== FUNCTION DETAILS ==================== -->
<a name="section-FunctionDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Function Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-FunctionDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="cleanFastaFile"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">cleanFastaFile</span>(<span class="sig-arg">fname</span>,
        <span class="sig-arg">targetfolder</span>=<span class="sig-default">''</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#cleanFastaFile">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Removes gaps from target FASTA file. Allows for BLAST input using this
  file. Creates new files for every sequence entry. Returns the filename(s)
  of the new file(s).</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="extractRandomFasta"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">extractRandomFasta</span>(<span class="sig-arg">fasta_file</span>,
        <span class="sig-arg">numRandom</span>,
        <span class="sig-arg">dest_file</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#extractRandomFasta">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Extracts a random number of sequences from a FASTA file. Writes to new
  FASTA.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="checkForFileCollision"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">checkForFileCollision</span>(<span class="sig-arg">path</span>,
        <span class="sig-arg">ext</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#checkForFileCollision">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Ensures a file is not being overwritten. Will recursively keep adding 
  a '0' to end of base filename until a unique filename is found. Returns 
  the path.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="system"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">system</span>(<span class="sig-arg">instruction</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#system">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Executes a system command line instruction. Returns list of stdout, 
  stderr.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="getAverageCorrelationCoefficient"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">getAverageCorrelationCoefficient</span>(<span class="sig-arg">len_struct</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#getAverageCorrelationCoefficient">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Gets the average correlation coefficient for randomly related 
  structure pairs given the length of a structure. See Betancourt and 
  Skolnick.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="radiusOfGyration"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">radiusOfGyration</span>(<span class="sig-arg">pdbFile</span>,
        <span class="sig-arg">chain</span>,
        <span class="sig-arg">pdb</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#radiusOfGyration">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Returns the approximate radius of gyration of a given PDB molecule. 
  Deprecated; now in PDBnet.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="tmscore"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">tmscore</span>(<span class="sig-arg">alignPDB</span>,
        <span class="sig-arg">alignFASTA</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#tmscore">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Returns the TMscore and associated P-value. See Xu &amp; Zhang, 
  &quot;How significant is a protein structure similarity with TM-score = 
  0.5?&quot;.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="rrmsd"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">rrmsd</span>(<span class="sig-arg">alignPDB</span>,
        <span class="sig-arg">alignFASTA</span>,
        <span class="sig-arg">rmsd</span>=<span class="sig-default">False</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#rrmsd">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Returns the rRMSD of an aligned pairwise PDB and FASTA file. See 
  Betancourt &amp; Skolnick, &quot;Universal Similarity Measure for 
  Comparison Protein Structures&quot;. Deprecated.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="rmsd"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">rmsd</span>(<span class="sig-arg">pdbFile1</span>,
        <span class="sig-arg">pdbFile2</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.homology-pysrc.html#rmsd">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Returns the RMSD of two PDB files. Requires: Pymol.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="src-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://web.cs.dal.ca/~safatli/">LabBlouinTools</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Mon Sep  8 11:33:36 2014
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
