<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>src.PDBnet.PDBstructure</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="src-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://web.cs.dal.ca/~safatli/">LabBlouinTools</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="src-module.html">Package&nbsp;src</a> ::
        <a href="src.PDBnet-module.html">Module&nbsp;PDBnet</a> ::
        Class&nbsp;PDBstructure
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="src.PDBnet.PDBstructure-class.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class PDBstructure</h1><p class="nomargin-top"><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure">source&nbsp;code</a></span></p>
<p>A PDB protein structure. Allows one to read in and parse PDB 
  files.</p>

<!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="__init__"></a><span class="summary-sig-name">__init__</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filein</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string"></code><code class="variable-quote">'</code></span>)</span><br />
      Construct a protein PDB (a collection of chains).</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.__init__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetChainNames"></a><span class="summary-sig-name">GetChainNames</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetChainNames">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetModelNames"></a><span class="summary-sig-name">GetModelNames</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetModelNames">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetChain"></a><span class="summary-sig-name">GetChain</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">ch</span>)</span><br />
      Get a chain by name.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetChain">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetModel"></a><span class="summary-sig-name">GetModel</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">mod</span>)</span><br />
      Get a model by name.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetModel">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#NewChain" class="summary-sig-name">NewChain</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">name</span>)</span><br />
      Construct and add a new chain by name to the PDB.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.NewChain">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#NewModel" class="summary-sig-name">NewModel</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">name</span>)</span><br />
      Construct and add a new model by name to the PDB.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.NewModel">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="AddChainToStructure"></a><span class="summary-sig-name">AddChainToStructure</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chainname</span>,
        <span class="summary-sig-arg">chain</span>)</span><br />
      Add a chain as a list of residues to the PDB.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddChainToStructure">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="AddModelToStructure"></a><span class="summary-sig-name">AddModelToStructure</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">modelname</span>,
        <span class="summary-sig-arg">model</span>)</span><br />
      Add a model as a list of residues to the PDB.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddModelToStructure">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#AddResidueToChain" class="summary-sig-name">AddResidueToChain</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chain</span>,
        <span class="summary-sig-arg">res</span>)</span><br />
      Add a residue to a chain.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddResidueToChain">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#AddResidueToModel" class="summary-sig-name">AddResidueToModel</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">model</span>,
        <span class="summary-sig-arg">res</span>)</span><br />
      Add a residue to a model.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddResidueToModel">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="AddRemark"></a><span class="summary-sig-name">AddRemark</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">remark</span>)</span><br />
      Add a remark (note/comment) to the structure/PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddRemark">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetRemarks"></a><span class="summary-sig-name">GetRemarks</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetRemarks">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="CheckComplete"></a><span class="summary-sig-name">CheckComplete</span>(<span class="summary-sig-arg">self</span>)</span><br />
      For every chain, check to see if every residue is complete (see 
      aa_list dictionary).</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.CheckComplete">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_pymol"></a><span class="summary-sig-name">_pymol</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure._pymol">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#view" class="summary-sig-name">view</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">istrajectory</span>=<span class="summary-sig-default">False</span>)</span><br />
      View the structure in a Pymol window.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.view">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="ViewStructure"></a><span class="summary-sig-name">ViewStructure</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ViewStructure">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="write"></a><span class="summary-sig-name">write</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Alias to WriteFile().</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.write">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#WriteChain" class="summary-sig-name">WriteChain</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">ch</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Write a single chain to a file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteChain">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#WriteModel" class="summary-sig-name">WriteModel</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">mod</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Write a single model to a file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteModel">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="WriteFile"></a><span class="summary-sig-name">WriteFile</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Write this PDB structure as a single PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteFile">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="read"></a><span class="summary-sig-name">read</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Alias for ReadFile().</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.read">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#ReadFile" class="summary-sig-name">ReadFile</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Read a PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ReadFile">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#ChainAsFASTA" class="summary-sig-name">ChainAsFASTA</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chain</span>)</span><br />
      Return the chain as a FASTA.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ChainAsFASTA">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#ModelAsFASTA" class="summary-sig-name">ModelAsFASTA</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">model</span>)</span><br />
      Return the model as a FASTA.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ModelAsFASTA">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#GetResidueAssociations" class="summary-sig-name">GetResidueAssociations</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Use an input FASTA file to determine associations between residues as
      they exist between 2+ chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetResidueAssociations">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#tmscore" class="summary-sig-name">tmscore</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>,
        <span class="summary-sig-arg">native</span>=<span class="summary-sig-default">None</span>)</span><br />
      Get the TMscore between two chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.tmscore">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#gdt" class="summary-sig-name">gdt</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">distcutoffs</span>=<span class="summary-sig-default"><code class="variable-group">[</code>1<code class="variable-op">, </code>2<code class="variable-op">, </code>4<code class="variable-op">, </code>8<code class="variable-group">]</code></span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Get the GDT score between two chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.gdt">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#rmsd" class="summary-sig-name">rmsd</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Get the RMSD between chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.rmsd">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#rrmsd" class="summary-sig-name">rrmsd</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Get the RRMSD between chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.rrmsd">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="RadiusOfGyration"></a><span class="summary-sig-name">RadiusOfGyration</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Acquire the radius of the gyration of the entire, or a portion of, 
      the PDB protein molecule.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.RadiusOfGyration">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetAllCentroid"></a><span class="summary-sig-name">GetAllCentroid</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chain</span>)</span><br />
      Populates the centroids of all residues.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetAllCentroid">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#IndexSeq" class="summary-sig-name">IndexSeq</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chain</span>,
        <span class="summary-sig-arg">fst</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Store in residues the correct index to the fasta.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.IndexSeq">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="GetFASTAIndices"></a><span class="summary-sig-name">GetFASTAIndices</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">thing</span>,
        <span class="summary-sig-arg">fst</span>)</span><br />
      Given a PDBchain, find 1-to-1 correspondances between it and a FASTA 
      sequence.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetFASTAIndices">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="IterAllResidues"></a><span class="summary-sig-name">IterAllResidues</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Produce an iterator to allow one to iterate over all possible 
      residues.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.IterAllResidues">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#WriteGM" class="summary-sig-name">WriteGM</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">gm</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>,
        <span class="summary-sig-arg">CA</span>=<span class="summary-sig-default">False</span>)</span><br />
      Write the information present in this PDB between multiple chains as 
      a Geometric Morphometric text file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteGM">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="src.PDBnet.PDBstructure-class.html#WriteLandmarks" class="summary-sig-name">WriteLandmarks</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fasta</span>,
        <span class="summary-sig-arg">lm</span>,
        <span class="summary-sig-arg">chains</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">ismodel</span>=<span class="summary-sig-default">False</span>)</span><br />
      Write the information present in this PDB between multiple chains as 
      a landmark text file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteLandmarks">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="Contacts"></a><span class="summary-sig-name">Contacts</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">chain</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">ALL</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">thres</span>=<span class="summary-sig-default">4.5</span>)</span><br />
      Compute the contact map of all chains or a chain.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.Contacts">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="WriteContacts"></a><span class="summary-sig-name">WriteContacts</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filename</span>)</span><br />
      Write contact map.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteContacts">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="__iter__"></a><span class="summary-sig-name">__iter__</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Returns all PDBchains as an iterator.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.__iter__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="__len__"></a><span class="summary-sig-name">__len__</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Returns the length of the protein.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.__len__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_remarksToString"></a><span class="summary-sig-name">_remarksToString</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure._remarksToString">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="__str__"></a><span class="summary-sig-name">__str__</span>(<span class="summary-sig-arg">self</span>)</span><br />
      As a string, outputs structure in the PDB format.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.__str__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="NewChain"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">NewChain</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">name</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.NewChain">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Construct and add a new chain by name to the PDB. Returns the 
  chain.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="NewModel"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">NewModel</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">name</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.NewModel">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Construct and add a new model by name to the PDB. Returns the 
  model.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="AddResidueToChain"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">AddResidueToChain</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">chain</span>,
        <span class="sig-arg">res</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddResidueToChain">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Add a residue to a chain. Deprecated; use chain class function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="AddResidueToModel"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">AddResidueToModel</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">model</span>,
        <span class="sig-arg">res</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.AddResidueToModel">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Add a residue to a model. Deprecated; use model class function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="view"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">view</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">istrajectory</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.view">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>View the structure in a Pymol window. Requires an installation of 
  Pymol.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="WriteChain"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">WriteChain</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">ch</span>,
        <span class="sig-arg">filename</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteChain">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Write a single chain to a file. Deprecated; use chain class 
  function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="WriteModel"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">WriteModel</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">mod</span>,
        <span class="sig-arg">filename</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteModel">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Write a single model to a file. Deprecated; use model class 
  function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="ReadFile"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">ReadFile</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">filename</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ReadFile">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Read a PDB file. Populate this PDBstructure.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="ChainAsFASTA"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">ChainAsFASTA</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">chain</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ChainAsFASTA">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Return the chain as a FASTA. Deprecated; use chain class function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="ModelAsFASTA"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">ModelAsFASTA</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">model</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.ModelAsFASTA">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Return the model as a FASTA. Deprecated; use chain class function.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="GetResidueAssociations"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">GetResidueAssociations</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.GetResidueAssociations">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Use an input FASTA file to determine associations between residues as 
  they exist between 2+ chains. Returns a mapping between chain and 
  sequence, a masking of gapped parts of the sequence(s), and a list of 
  aligned positions.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="tmscore"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">tmscore</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>,
        <span class="sig-arg">native</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.tmscore">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the TMscore between two chains. Requires a FASTA alignment and a 
  value for the length of the native structure (e.g., for a pairwise 
  alignment, the length of the structure used as a reference before 
  alignment was done). The latter is computed by assuming the first of both
  provided chains is the native structure; otherwise, uses a provided chain
  name (native input).</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="gdt"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">gdt</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">distcutoffs</span>=<span class="sig-default"><code class="variable-group">[</code>1<code class="variable-op">, </code>2<code class="variable-op">, </code>4<code class="variable-op">, </code>8<code class="variable-group">]</code></span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.gdt">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the GDT score between two chains. Requires a FASTA alignment.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="rmsd"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">rmsd</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.rmsd">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the RMSD between chains. Requires a FASTA alignment.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="rrmsd"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">rrmsd</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.rrmsd">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the RRMSD between chains. Requires a FASTA alignment. See 
  Betancourt &amp; Skolnick, &quot;Universal Similarity Measure for 
  Comparison Protein Structures&quot;.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="IndexSeq"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">IndexSeq</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">chain</span>,
        <span class="sig-arg">fst</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.IndexSeq">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Store in residues the correct index to the fasta. Requires a 1-to-1 
  correspondance at least a portion of the way through. Deprecated; use 
  GetFASTAIndices().</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="WriteGM"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">WriteGM</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">gm</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>,
        <span class="sig-arg">CA</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteGM">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Write the information present in this PDB between multiple chains as a
  Geometric Morphometric text file. This file will be formatted such that 
  individual lines correspond to chains and semi-colons separate the 
  (x,y,z) coordinates between all homologous residue positions. Requires a 
  FASTA alignment. Options include using alpha-carbon positions. By 
  default, uses centroids of residues.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="WriteLandmarks"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">WriteLandmarks</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fasta</span>,
        <span class="sig-arg">lm</span>,
        <span class="sig-arg">chains</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">ismodel</span>=<span class="sig-default">False</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="src.PDBnet-pysrc.html#PDBstructure.WriteLandmarks">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Write the information present in this PDB between multiple chains as a
  landmark text file. This file will be formatted such that the file is 
  partitioned in sections starting with chain names and individual lines in
  these correspond to homologous residue positions denoted by homologous 
  position, residue number, and residue name tab-delimited. Requires a 
  FASTA file.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="src-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://web.cs.dal.ca/~safatli/">LabBlouinTools</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Mon Sep  8 11:33:36 2014
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
